src/ports/postgres/modules/utilities/pivot.py_in

```diff
@@ -58,66 +61,256 @@ def pivot(schema_madlib, source_table, out_table,
-    case_str = ("{agg}("
-                "CASE WHEN \"{{pcol}}\" = '{{value}}' THEN {pval} ELSE NULL END"
-                ")".
-                format(agg=aggregate_func,
-                       pval=pval_no_quotes))
-    sql_list.append(
-        ", " +
-        # Assign the name of the new column
-        ', '.join("{case_str} as \"{{pcol}}_{{value}}\"".
-                  format(case_str=case_str).
-                  format(pcol=pcol_no_quotes, value=str(value))
-                  for value in distinct_values if value is not None))
+    # Note that the distinct values are appended in order of users list
+    # This ordering is important when we access pivot_comb entries
+    array_agg_str = ', '.join("array_agg(DISTINCT {pcol}) AS {pcol}_values".
+        format(pcol=pcol) for pcol in pcols)
+
+    null_str = ""
+    if keep_null:
+        # Create an additional column for every pivot column
+        # If there is a null value, this column will get True
+        null_str = ","+', '.join(
+            "bool_or(CASE WHEN {pcol} IS NULL THEN True END)"
+            "AS {pcol}_isnull".format(pcol=pcol) for pcol in pcols)
+
+    distinct_values = plpy.execute("SELECT {0} {1} FROM {2}".
+        format(array_agg_str, null_str, source_table))
+
+    # Collect the distinct values for every pivot column
+    pcol_distinct_values = {}
+    pcol_max_length = 0
+    for pcol in pcols:
+        # Read the distinct values for this pcol
+        pcol_tmp = [item for item in distinct_values[0][pcol+"_values"]]
+        # Remove null values if keep null is not true
+        if not keep_null:
+            pcol_tmp = [x for x in pcol_tmp if x is not None]
+        elif distinct_values[0][pcol+"_isnull"] and None not in pcol_tmp:
+            pcol_tmp.append(None)
+
+        pcol_distinct_values[pcol]=sorted(pcol_tmp)
+        # Max length of the string that pcol values can create +
+        # length of pcol + 1 (for _ character)
+        pcol_max_length+=max([len(str(item)) for item in pcol_tmp])+len(pcol)+1
+
+    # Create the combination of every possible pivot column
+    # Assume piv and piv2 are pivot columns. piv=(1,2) and piv2=(3,4,5)
+    # pivot_comb = ((1,3),(1,4),(1,5),(2,3),(2,4),(2,5))
+    pivot_comb = list(itertools.product(*([pcol_distinct_values[pcol]
+        for pcol in pcols])))
+    #Prepare the wrapper for fill value
+    fill_str_begin = ""
+    fill_str_end = ""
+    if fill_value is not None:
+        fill_str_begin = " COALESCE("
+        fill_str_end = ", "+fill_value+" ) "
+
+    # Check the max possible length of a output column name
+    # If it is over 63 (psql upper limit) create table lookup
+    for pval in pvals:
+
+        col_name_len = pcol_max_length+len(pval)+1
+        try:
+            # If user specifies a list of aggregates for a value column
+            # Every value column has to have an entry
+            agg_func = agg_dict[pval] if len(agg_dict) > 1 else \
```

Comment:
I added the default to the dictionary to keep track of one less variable name and keeping all of aggregate functions in the same list seemed like an acceptable idea. 

Yes that is the behavior. I can change the behavior to assume the default function (avg) for the missing value columns but that will bring a new issue where the user tries to define the default as well as a list.